src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1043,139 +1068,644 @@ private StringConcatFactory() {
+            }
+            return changed ? MethodType.methodType(args.returnType(), paramTypes) : args;
+        }
 
-        static final Set<Lookup.ClassOption> SET_OF_STRONG = Set.of(STRONG);
+        /**
+         * Construct the MethodType of the prepend method, The parameters only support 5 types:
+         * int/long/char/boolean/String. Not int/long/char/boolean type, use String type<p>
+         *
+         * The following is an example of the generated target code:
+         * <blockquote><pre>
+         *  int prepend(int length, byte coder, byte[] buff,  String[] constants
+         *      int arg0, long arg1, boolean arg2, char arg3, String arg5)
+         * </pre></blockquote>
+         */
+        private static MethodTypeDesc prependArgs(MethodType concatArgs) {
+            int parameterCount = concatArgs.parameterCount();
+            var paramTypes = new ClassDesc[parameterCount + 4];
+            paramTypes[0] = CD_int;          // length
+            paramTypes[1] = CD_byte;         // coder
+            paramTypes[2] = CD_Array_byte;   // buff
+            paramTypes[3] = CD_Array_String; // constants
+
+            for (int i = 0; i < parameterCount; i++) {
+                var cl = concatArgs.parameterType(i);
+                if (cl != String.class && needStringOf(cl)) {
+                    cl = String.class;
+                }
+                paramTypes[i + 4] = ConstantUtils.classDesc(cl);
+            }
+            return MethodTypeDesc.of(CD_int, paramTypes);
+        }
 
-        private SimpleStringBuilderStrategy() {
-            // no instantiation
+        /**
+         * Construct the MethodType of the coder method,
+         * The first parameter is the initialized coder, Only parameter types that can be UTF16 are added.
+         */
+        private static MethodTypeDesc coderArgs(MethodType concatArgs) {
+            int parameterCount = concatArgs.parameterCount();
+            List<ClassDesc> paramTypes = new ArrayList<>();
+            paramTypes.add(CD_int); // init coder
+            for (int i = 0; i < parameterCount; i++) {
+                var cl = concatArgs.parameterType(i);
+                if (maybeUTF16(cl)) {
+                    paramTypes.add(cl == char.class ? CD_char : CD_String);
+                }
+            }
+            return MethodTypeDesc.of(CD_int, paramTypes);
+        }
+
+        /**
+         * Construct the MethodType of the length method,
+         * The first parameter is the initialized length
+         */
+        private static MethodTypeDesc lengthArgs(MethodType concatArgs) {
+            int parameterCount = concatArgs.parameterCount();
+            var paramTypes = new ClassDesc[parameterCount + 1];
+            paramTypes[0] = CD_int; // init long
+            for (int i = 0; i < parameterCount; i++) {
+                var cl = concatArgs.parameterType(i);
+                paramTypes[i + 1] = needString(cl) ? CD_String : ConstantUtils.classDesc(cl);
```

Comment:
```suggestion
                paramTypes[i + 1] = needStringOf(cl) ? CD_String : ConstantUtils.classDesc(cl);
```

Typo in my previous suggestion